### PR TITLE
Convert Ingress to Gateway

### PR DESCRIPTION
**Pull request by esnible**
An Ingress is now converted into a Gateway + VirtualService.

### CODE REVIEW

File: istioctl/cmd/istioctl/convert/testdata/v1alpha3/myservice-gateway.yaml.golden [hunk 1]
```diff
@@ -0,0 +1,51 @@
+apiVersion: networking.istio.io/v1alpha3
+kind: Gateway
+metadata:
+  creationTimestamp: null
+  name: istio-autogenerated-k8s-ingress
```
Comment (rshriram) [inline istioctl/cmd/istioctl/convert/testdata/v1alpha3/myservice-gateway.yaml.golden]: Is this the name we set internally ? I think it is but double check

File: istioctl/cmd/istioctl/convert/testdata/v1alpha3/myservice-gateway.yaml.golden [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  namespace: istio-system
+spec:
+  gateways:
+  - istio-autogenerated-k8s-ingress
+  hosts:
+  - '*'
+  http:
+  - match:
+    - authority:
+        regex: ""
```
Comment (rshriram) [inline istioctl/cmd/istioctl/convert/testdata/v1alpha3/myservice-gateway.yaml.golden]: This is going to cause error

File: istioctl/cmd/istioctl/convert/testdata/v1alpha3/myservice-gateway.yaml.golden [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+      uri:
+        regex: /foo/.*
```
Comment (rshriram) [inline istioctl/cmd/istioctl/convert/testdata/v1alpha3/myservice-gateway.yaml.golden]: Should be prefix /foo/

File: istioctl/pkg/convert/ingress.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+// IstioIngresses converts K8s extensions/v1beta1 Ingresses with Istio rules to v1alpha3 gateway and virtual service
+func IstioIngresses(ingresses []*v1beta1.Ingress, domainSuffix string) ([]model.Config, error) {
+	
+	out := make([]model.Config, 0)
+
+	for _, ingrezz := range ingresses {
+		gateway, virtualService := ingress.ConvertIngressV1alpha3(*ingrezz, domainSuffix)
+		out = append(out, gateway)
+		out = append(out, virtualService)
+	}
```
Comment (rshriram) [inline istioctl/pkg/convert/ingress.go]: This logic is wrong. You need only one gateway for all of istio ingress with wildcard host *, and ports 80/443. And then only one virtual service pinned to the gateway. 

What you need to do here is 
For each ingress, get the gateway.server(443) and append it to the Uber gateway.servers if and only if server has non wildcard host (ie not *). If it has only wildcard, append only once (cant have multiple tls servers on wildcard host).

For virtual services, append all non empty match blocks to the Uber virtual service (where non empty is a match that has authority or path match - not prefix /).

Comment (ZackButcher) [reply istioctl/pkg/convert/ingress.go]: I don't think we want this behavior long term (just one Gateway, I mean). I'm not 100% sure I'll need to change it to get all of the ingress tests passing, but in general we should support multiple 80/443 Gateways for ingress. I don't see a compelling reason we can't just blindly create a duplicate gateway per ingress.

File: istioctl/cmd/istioctl/convert/cmd.go [hunk 5]
```diff
@@ -142,25 +149,57 @@ func convertConfigs(readers []io.Reader, writer io.Writer) error {
...
 		if err != nil {
-			return nil, err
+			return nil, nil, err
+		}
+
+		recognized := 0
+		for _, nonIstio := range kinds {
+			if nonIstio.Kind == "Ingress" &&
+				nonIstio.APIVersion == "extensions/v1beta1" &&
+				nonIstio.ObjectMeta.Annotations["kubernetes.io/ingress.class"] == "istio" {
```
Comment (ZackButcher) [inline istioctl/cmd/istioctl/convert/cmd.go]: I don't think we care about ingress class annotation given this is used in a command line context.

File: istioctl/cmd/istioctl/convert/testdata/v1alpha3/merged-gateway.yaml.golden [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    port:
+      name: http-ingress-80
+      number: 80
+      protocol: HTTP
+  - hosts:
+    - '*'
+    port:
+      name: http-ingress-80
+      number: 80
+      protocol: HTTP
```
Comment (rshriram) [inline istioctl/cmd/istioctl/convert/testdata/v1alpha3/merged-gateway.yaml.golden]: this does not look right.. duplicates

File: istioctl/cmd/istioctl/convert/testdata/v1alpha3/merged-gateway.yaml.golden [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+kind: VirtualService
+metadata:
+  creationTimestamp: null
+  name: simple-ingress-istio-autogenerated-k8s-ingress
+  namespace: istio-system
+spec:
+  gateways:
+  - simple-ingress-istio-autogenerated-k8s-ingress
+  hosts:
+  - '*'
```
Comment (rshriram) [inline istioctl/cmd/istioctl/convert/testdata/v1alpha3/merged-gateway.yaml.golden]: cannot have multiple virtual services for same host

File: pilot/pkg/config/kube/ingress/conversion.go [hunk 8]
```diff
@@ -326,3 +318,31 @@ func shouldProcessIngress(mesh *meshconfig.MeshConfig, ingress *v1beta1.Ingress)
...
+	// Replace "foo" with exact match
+	count := strings.Count(s, ".*")
+	if count == 0 {
+		return &networking.StringMatch{
+			MatchType: &networking.StringMatch_Exact{Exact: s},
+		}
+	}
+
+	// Replace "foo.*bar.*" and "foo.*bar" with regex match
+	if count > 1 || !strings.HasSuffix(s, ".*") {
```
Comment (rshriram) [inline pilot/pkg/config/kube/ingress/conversion.go]: This is not the right way to check for regex strings. See proxy envoy v1 ingress.go. There is a go library thing that can check if a string is a regex or not.

Comment (rshriram) [reply pilot/pkg/config/kube/ingress/conversion.go]: Actually can you just remove this entire piece ?

File: pilot/pkg/config/kube/ingress/conversion.go [hunk 9]
```diff
@@ -326,3 +318,31 @@ func shouldProcessIngress(mesh *meshconfig.MeshConfig, ingress *v1beta1.Ingress)
...
+				Regex: s,
+			},
+		}
+	}
+
+	// Replace "foo.*" with a Prefix match
+	return &networking.StringMatch{
+		MatchType: &networking.StringMatch_Prefix{Prefix: strings.TrimSuffix(s, ".*")},
+	}
+}
```
Comment (rshriram) [inline pilot/pkg/config/kube/ingress/conversion.go]: We just need exact and prefix match. Ignore regex match for now.

**Comment by esnible** [comment]
PR https://github.com/istio/istio/pull/5411 was merged instead of this one.

**Pull request closed without merging**